apm-agent-plugins/apm-jdk-httpserver-plugin/src/main/java/co/elastic/apm/agent/httpserver/HttpHandlerAdvice.java

```diff
@@ -0,0 +1,167 @@
+            Headers headers = exchange.getRequestHeaders();
+            if (headers != null) {
+                for (Map.Entry<String, List<String>> header : headers.entrySet()) {
+                    if ("Cookie".equalsIgnoreCase(header.getKey())) {
+                        for (String[] cookie : CookieHelper.getCookies(header.getValue())) {
+                            request.addCookie(cookie[0], cookie[1]);
+                        }
+                    } else {
+                        request.addHeader(header.getKey(), Collections.enumeration(header.getValue()));
+                    }
+                }
+            }
+        }
+
+        return transaction.activate();
+    }
+
+    public static boolean isExcluded(WebConfiguration webConfiguration, String requestPath, @Nullable String userAgentHeader) {
+        WildcardMatcher excludeUrlMatcher = WildcardMatcher.anyMatch(webConfiguration.getIgnoreUrls(), requestPath);
+        if (excludeUrlMatcher != null) {
+            logger.debug("Not tracing this request as the URL {} is ignored by the matcher {}", requestPath, excludeUrlMatcher);
+        }
+        WildcardMatcher excludeAgentMatcher = userAgentHeader != null ? WildcardMatcher.anyMatch(webConfiguration.getIgnoreUserAgents(), userAgentHeader) : null;
+        if (excludeAgentMatcher != null) {
+            logger.debug("Not tracing this request as the User-Agent {} is ignored by the matcher {}", userAgentHeader, excludeAgentMatcher);
+        }
+        boolean isExcluded = excludeUrlMatcher != null || excludeAgentMatcher != null;
+        if (!isExcluded) {
+            logger.trace("No matcher found for excluding this request with request-path: {} and User-Agent: {}", requestPath, userAgentHeader);
+        }
+        return isExcluded;
+    }
+
+    private static String getHostname(HttpExchange exchange) {
+        List<String> hostHeader = exchange.getRequestHeaders().get("Host");
+        if (hostHeader != null) {
+            String hostname = hostHeader.get(0);
+
+            int idx = hostname.lastIndexOf(':');
+            if (idx > 0) {
+                String port = String.valueOf(exchange.getLocalAddress().getPort());
+                if (idx + 1 + port.length() == hostname.length() && hostname.endsWith(port)) {
+                    hostname = hostname.substring(0, idx); // remove port from hostname
+                }
+            }
+
+            return hostname;
+        }
+        return exchange.getLocalAddress().getAddress().getHostName();
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+    public static void onExitHandle(@Advice.Argument(0) HttpExchange exchange, @Advice.Enter @Nullable Object transactionOrNull, @Advice.Thrown @Nullable Throwable t) {
+        ElasticApmTracer tracer = GlobalTracer.getTracerImpl();
+        if (tracer == null || transactionOrNull == null) {
```

Comment:
No need to check for tracer state as `transactionOrNull` will be `null` when `GlobalTracer.getTracerImpl()` has already returned `null` in `onEnterHandle(...)`

Reply:
Done